packages/pigeon/lib/cpp_generator.dart

```diff
@@ -1417,6 +1478,170 @@ void _writeSystemHeaderIncludeBlock(Indent indent, List<String> headers) {
   }
 }
 
+enum _FunctionOutputType { declaration, definition }
+
+/// Writes a function declaration or definition to [indent].
+///
+/// If [parameters] are given, each should be a string of the form 'type name'.
+void _writeFunction(
+  Indent indent,
+  _FunctionOutputType type, {
+  required String name,
+  String? returnType,
+  String? scope,
+  List<String> parameters = const <String>[],
+  List<String> startingAnnotations = const <String>[],
+  List<String> trailingAnnotations = const <String>[],
+  List<String> initializers = const <String>[],
+  void Function()? body,
+}) {
+  assert(body == null || type == _FunctionOutputType.definition);
+
+  // Set the initial indentation.
+  indent.write('');
+  // Write any starting annotations (e.g., 'static').
+  for (final String annotation in startingAnnotations) {
+    indent.add('$annotation ');
+  }
+  // Write the signature.
+  if (returnType != null) {
+    indent.add('$returnType ');
+  }
+  if (scope != null) {
+    indent.add('$scope::');
+  }
+  indent.add(name);
+  // Write the parameters.
+  if (parameters.isEmpty) {
+    indent.add('()');
+  } else if (parameters.length == 1) {
+    indent.add('(${parameters.first})');
+  } else {
+    indent.addScoped('(', null, () {
+      enumerate(parameters, (int index, final String param) {
+        if (index == parameters.length - 1) {
+          indent.write('$param)');
+        } else {
+          indent.writeln('$param,');
+        }
+      });
+    }, addTrailingNewline: false);
+  }
+  // Write any trailing annotations (e.g., 'const').
+  for (final String annotation in trailingAnnotations) {
+    indent.add(' $annotation');
+  }
+  // Write the initializer list, if any.
+  if (initializers.isNotEmpty) {
+    indent.newln();
+    indent.write(' : ');
+    // The first item goes on the same line as the ":", the rest go on their
+    // own lines indented two extra levels, with no comma or newline after the
+    // last one. The easiest way to express the special casing of the first and
+    // last is with a join+format.
+    indent.format(initializers.join(',\n\t\t'),
```

Comment:
This one I still found harder to understand without using format, per the comment, but I could be persuaded.